### PR TITLE
web: make HTTP field names case-insensitive

### PR DESCRIPTION
**Pull request by akshayjshah**
I'm opening this PR to (hopefully!) stimulate a discussion. In brief, I'd like to amend the gRPC-Web protocol docs to encourage implementations to follow HTTP semantics and compare HTTP field names case-insensitively.

The gRPC-Web specification is a nicely-designed way for proxies to expose standard HTTP/2 gRPC servers to clients using less tightly-controlled HTTP stacks, such as web browsers. To serve that goal, it seems valuable to have the gRPC-Web specification follow [RFC 9110 (HTTP Semantics)](https://www.rfc-editor.org/rfc/rfc9110.html#name-field-names). Like previous RFCs, 9110 specifies that "field names are case-insensitive." However, the current gRPC-Web specification requires that servers and proxies "use lower-case header/trailer names" on the wire. In principle, mandating casing on the wire is normal for HTTP/2 and fine (if unusual) for HTTP/1.1; however, it encourages implementations to violate HTTP semantics and require lower-case names when _reading_ headers and trailers.

I'd like to loosen the gRPC-Web specification to permit any casing on the wire for HTTP/1.1. I'd also like to emphasize that gRPC-Web implementations ought to follow standard HTTP semantics when _reading_ fields and compare names case-insensitively.  Implementations that can't treat names case-insensitively without breaking backward compatibility should instead normalize field names to lowercase. Among the Google-maintained gRPC implementations, at least `grpc-go` and `grpc-java` already compare names case-insensitively (even though they're HTTP/2-only). `grpc-dart` does the opposite and compares names case-sensitively. `grpc-web` is sometimes case-insensitive (when reading `grpc-status` and `grpc-message` from trailers-only responses) and sometimes case-sensitive (when hand-parsing a block of length-prefixed trailers).

The proposed amendment does not affect the correctness of Envoy (which may continue to use lower-case field names). It partially affects `grpc-web`, which would require a small patch to always normalize names. (Both patched and unpatched versions of `grpc-web` would work with Envoy.) `grpc-dart` would need to either begin treating field names case-insensitively or normalize names, depending on what's possible in Dart without breaking backward compatibility.

Relates to https://github.com/improbable-eng/grpc-web/issues/228, https://github.com/bufbuild/connect-go/issues/453/, and https://github.com/grpc/grpc-dart/issues/594.

### CODE REVIEW

**Comment by akshayjshah** [comment]
@wenbozhu I've amended the diff here - it's hopefully a bit clearer now. Appreciate you taking a look!

The mergeability check is currently failing because of missing labels. I'm not sure which of the labels is appropriate here, or if this change should be included in the changelog. Happy to take your advice :)

**Comment by akshayjshah** [comment]
@wenbozhu Any thoughts on the contents of this PR?

File: doc/PROTOCOL-WEB.md
Comment (wenbozhu) [inline doc/PROTOCOL-WEB.md:51]: How about 

2. header names may not be lower-case over HTTP/1.1; and trailers are not supported

**Review by wenbozhu — COMMENTED** [review]
Thanks for the suggestion.

File: doc/PROTOCOL-WEB.md
Comment (akshayjshah) [reply doc/PROTOCOL-WEB.md:51]: I believe the current language is referring to the block of trailers encoded in the last length-prefixed message in the stream, not HTTP trailers. Those trailers aren't visible to any generic HTTP systems (proxies, middleware, etc.) so I think it's fine to mandate lower-case field names on the wire.

How about this?

> HTTP header and trailer names may not be lower-case over HTTP/1.1; trailers encoded in the final length-prefixed message of the stream must use lower-case names

**Comment by akshayjshah** [comment]
@wenbozhu Thanks for the feedback - I've updated the PR!

**Review by wenbozhu — APPROVED** [review]

**Comment by akshayjshah** [comment]
:tada: Thanks for the stamp! What do I need to do to get CI passing and merge this in?

**Comment by sampajano** [comment]
@akshayjshah Thanks for the proposal and the change! I wonder if would provide a patch for https://github.com/grpc/grpc-web too, if it's within your interest? If not, we can do a follow up change there. Thanks! :)

**Comment by akshayjshah** [comment]
@sampajano Unfortunately, I'm probably not the right person to make a patch to any JS project 🤦🏽‍♂️

**Comment by sampajano** [comment]
@akshayjshah aha ok! perfectly fine! Thanks for considering! :)

I'll make a follow-up change once this change is submitted then! thanks!

**Comment by akshayjshah** [comment]
@wenbozhu or @sampajano - is there anything else I need to do to get this merged in? I'm not sure how to add the appropriate label to get the `Mergeable` CI check passing, and none of the other checks have run.

**Comment by sampajano** [comment]
> @wenbozhu or @sampajano - is there anything else I need to do to get this merged in? I'm not sure how to add the appropriate label to get the `Mergeable` CI check passing, and none of the other checks have run.

@akshayjshah Hi! I've added the missing label so i think the "Mergeable" CI should pass.

I'd wait for @wenbozhu for approval since there's a comment from him. I'll ping the thread too!

File: doc/PROTOCOL-WEB.md
Comment (sampajano) [reply doc/PROTOCOL-WEB.md:51]: @wenbozhu Could you take a look and approve the PR if the updated text looks good? Thanks!

**Comment by akshayjshah** [comment]
Thanks for merging this in, @sampajano! 😻

**Pull request merged**
